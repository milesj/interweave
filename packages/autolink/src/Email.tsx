import React from 'react';
import Link from './Link';
import { EmailProps } from './types';

export default function Email({ children, email, ...props }: EmailProps) {
  return (
    <Link {...props} href={`mailto:${email}`}>
      {children}
    </Link>
  );
}
