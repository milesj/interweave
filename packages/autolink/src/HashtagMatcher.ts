import React from 'react';
import { ChildrenNode, Matcher, MatchResponse, Node } from 'interweave';
import { HASHTAG_PATTERN } from './constants';
import { Hashtag } from './Hashtag';
import { HashtagProps } from './types';

export type HashtagMatch = Pick<HashtagProps, 'hashtag'>;

export class HashtagMatcher extends Matcher<HashtagProps> {
	replaceWith(children: ChildrenNode, props: HashtagProps): Node {
		return React.createElement(Hashtag, props, children);
	}

	asTag(): string {
		return 'a';
	}

	match(string: string): MatchResponse<HashtagMatch> | null {
		return this.doMatch(string, HASHTAG_PATTERN, (matches) => ({
			hashtag: matches[0],
		}));
	}
}
