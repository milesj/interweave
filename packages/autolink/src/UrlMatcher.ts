import React from 'react';
import { ChildrenNode, Matcher, MatchResponse, Node } from 'interweave';
import { EMAIL_DISTINCT_PATTERN, TOP_LEVEL_TLDS, URL_PATTERN } from './constants';
import { UrlMatcherOptions, UrlProps } from './types';
import { Url } from './Url';

export type UrlMatch = Pick<UrlProps, 'url' | 'urlParts'>;

export class UrlMatcher extends Matcher<UrlProps, UrlMatcherOptions> {
	constructor(
		name: string,
		options?: UrlMatcherOptions,
		factory?: React.ComponentType<UrlProps> | null,
	) {
		super(
			name,
			{
				customTLDs: [],
				validateTLD: true,
				...options,
			},
			factory,
		);
	}

	replaceWith(children: ChildrenNode, props: UrlProps): Node {
		return React.createElement(Url, props, children);
	}

	asTag(): string {
		return 'a';
	}

	match(string: string): MatchResponse<UrlMatch> | null {
		const response = this.doMatch(string, URL_PATTERN, this.handleMatches);

		// False positives with URL auth scheme
		if (response?.match.match(EMAIL_DISTINCT_PATTERN)) {
			response.valid = false;
		}

		if (response?.valid && this.options.validateTLD) {
			const { host } = (response.urlParts as unknown) as UrlProps['urlParts'];
			const validList = [...TOP_LEVEL_TLDS, ...(this.options.customTLDs ?? [])];
			const tld = host.slice(host.lastIndexOf('.') + 1).toLowerCase();

			if (!validList.includes(tld)) {
				return null;
			}
		}

		return response;
	}

	/**
	 * Package the matched response.
	 */
	handleMatches(matches: string[]): UrlMatch {
		return {
			url: matches[0],
			urlParts: {
				auth: matches[2] ? matches[2].slice(0, -1) : '',
				fragment: matches[7] || '',
				host: matches[3],
				path: matches[5] || '',
				port: matches[4] ? matches[4] : '',
				query: matches[6] || '',
				scheme: matches[1] ? matches[1].replace('://', '') : 'http',
			},
		};
	}
}
