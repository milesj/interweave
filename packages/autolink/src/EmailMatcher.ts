import React from 'react';
import { ChildrenNode, Matcher, MatchResponse, Node } from 'interweave';
import { EMAIL_PATTERN } from './constants';
import { Email } from './Email';
import { EmailProps } from './types';

export type EmailMatch = Pick<EmailProps, 'email' | 'emailParts'>;

export class EmailMatcher extends Matcher<EmailProps> {
	replaceWith(children: ChildrenNode, props: EmailProps): Node {
		return React.createElement(Email, props, children);
	}

	asTag(): string {
		return 'a';
	}

	match(string: string): MatchResponse<EmailMatch> | null {
		return this.doMatch(string, EMAIL_PATTERN, (matches) => ({
			email: matches[0],
			emailParts: {
				host: matches[2],
				username: matches[1],
			},
		}));
	}
}
