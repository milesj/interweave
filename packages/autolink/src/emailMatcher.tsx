import React from 'react';
import { createMatcher } from 'interweave';
import Email from './Email';
import { EMAIL_PATTERN } from './constants';
import { EmailMatch } from './types';

export default createMatcher<EmailMatch, object>(
  EMAIL_PATTERN,
  ({ email }, props, children) => <Email email={email}>{children}</Email>,
  {
    onMatch: ({ matches }) => ({
      email: matches[0],
      parts: {
        host: matches[2],
        username: matches[1],
      },
    }),
    tagName: 'a',
  },
);
