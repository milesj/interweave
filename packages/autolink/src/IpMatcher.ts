import React from 'react';
import { MatchResponse } from 'interweave';
import { IP_PATTERN } from './constants';
import { UrlMatcherOptions, UrlProps } from './types';
import UrlMatcher, { UrlMatch } from './UrlMatcher';

export default class IpMatcher extends UrlMatcher {
	constructor(
		name: string,
		options?: UrlMatcherOptions,
		factory?: React.ComponentType<UrlProps> | null,
	) {
		super(
			name,
			{
				...options,
				// IPs dont have TLDs
				validateTLD: false,
			},
			factory,
		);
	}

	override match(string: string): MatchResponse<UrlMatch> | null {
		return this.doMatch(string, IP_PATTERN, this.handleMatches);
	}
}
