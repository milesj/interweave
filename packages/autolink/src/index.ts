/**
 * @copyright   2016-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import emailMatcher from './emailMatcher';
import hashtagMatcher from './hashtagMatcher';
import ipMatcher from './ipMatcher';
import urlMatcher from './urlMatcher';
import Email from './Email';
import Hashtag from './Hashtag';
import Link from './Link';
import Url from './Url';

export { emailMatcher, hashtagMatcher, ipMatcher, urlMatcher, Email, Hashtag, Link, Url };

export * from './constants';
export * from './types';
