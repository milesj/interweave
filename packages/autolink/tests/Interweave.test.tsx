import React from 'react';
import ReactDOMServer from 'react-dom/server';
import Interweave, { InterweaveProps } from 'interweave';
import { polyfill } from 'interweave-ssr';
import { render } from 'rut-dom';
import { EmailMatcher } from '../src/EmailMatcher';
import { HashtagMatcher } from '../src/HashtagMatcher';
import { IpMatcher } from '../src/IpMatcher';
import { UrlMatcher } from '../src/UrlMatcher';

describe('Interweave (with autolinking)', () => {
	it('renders large blocks of text with all matchers', () => {
		const result = render<InterweaveProps>(
			<Interweave
				content={`Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec massa lorem, mollis non commodo quis, ultricies at elit. email@domain.com. Aliquam a arcu porttitor, aliquam eros sed, convallis massa. Nunc vitae vehicula quam, in feugiat ligula. #interweave Donec eu sem non nibh condimentum luctus. Vivamus pharetra feugiat blandit. Vestibulum neque velit, semper id vestibulum id, viverra a felis. Integer convallis in orci nec bibendum. Ut consequat posuere metus, www.domain.com.

Curabitur lectus odio, tempus quis velit vitae, cursus sagittis nulla. Maecenas sem nulla, tempor nec risus nec, ultricies ultricies magna. https://127.0.0.1/foo Nulla malesuada lacinia libero non mollis. Curabitur id lacus id dolor vestibulum ornare quis a nisi (http://domain.com/some/path?with=query). Pellentesque ac finibus mauris. Sed eu luctus diam. Quisque porta lectus in turpis imperdiet dapibus.

#blessed #interweave #milesj`}
				matchers={[
					new EmailMatcher('email'),
					new HashtagMatcher('hashtag'),
					new IpMatcher('ip'),
					new UrlMatcher('url'),
				]}
				tagName="div"
			/>,
		);

		expect(result).toMatchSnapshot();
	});

	it('renders HTML text with all matchers', () => {
		const result = render<InterweaveProps>(
			<Interweave
				content={`<h1>Lorem ipsum dolor sit amet</h1>

<p><b>Consectetur adipiscing elit.</b> Donec massa lorem, mollis non commodo quis, ultricies at elit. email@domain.com. Aliquam a arcu porttitor, aliquam eros sed, convallis massa. Nunc vitae vehicula quam, in feugiat ligula. #interweave Donec eu sem non nibh condimentum luctus. Vivamus pharetra feugiat blandit. Vestibulum neque velit, semper id vestibulum id, viverra a felis. Integer convallis in orci nec bibendum. Ut consequat posuere metus, <a href="www.domain.com">www.domain.com</a>.</p>

<div>Curabitur lectus odio, <em>tempus quis velit vitae, cursus sagittis nulla</em>. Maecenas sem nulla, tempor nec risus nec, ultricies ultricies magna. https://127.0.0.1/foo Nulla malesuada lacinia libero non mollis. Curabitur id lacus id dolor vestibulum ornare quis a nisi (http://domain.com/some/path?with=query). Pellentesque ac finibus mauris. Sed eu luctus diam. Quisque porta lectus in turpis imperdiet dapibus.</div>

<section>#blessed #interweave #milesj</section>`}
				matchers={[
					new EmailMatcher('email'),
					new HashtagMatcher('hashtag'),
					new IpMatcher('ip'),
					new UrlMatcher('url'),
				]}
				tagName="div"
			/>,
		);

		expect(result).toMatchSnapshot();
	});

	it('doesnt render anchor links within anchor links', () => {
		const result = render<InterweaveProps>(
			<Interweave
				content={`- https://127.0.0.1/foo
- <a href="www.domain.com">www.domain.com</a>
- (http://domain.com/some/path?with=query)
- <a href="http://domain.com">This text should stay</a>`}
				matchers={[
					new EmailMatcher('email'),
					new HashtagMatcher('hashtag'),
					new IpMatcher('ip'),
					new UrlMatcher('url'),
				]}
				tagName="div"
			/>,
		);

		expect(result).toMatchSnapshot();
	});

	describe('server side rendering', () => {
		let implSpy: jest.SpyInstance;

		beforeEach(() => {
			polyfill();

			implSpy = jest.spyOn(document.implementation, 'createHTMLDocument');
		});

		afterEach(() => {
			global.INTERWEAVE_SSR_POLYFILL = undefined;

			implSpy.mockRestore();
		});

		it('supports complex content', () => {
			const actual = ReactDOMServer.renderToStaticMarkup(
				<Interweave
					content={`Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec massa lorem, mollis non commodo quis, ultricies at elit. email@domain.com. Aliquam a arcu porttitor, aliquam eros sed, convallis massa. Nunc vitae vehicula quam, in feugiat ligula. #interweave Donec eu sem non nibh condimentum luctus. Vivamus pharetra feugiat blandit. Vestibulum neque velit, semper id vestibulum id, viverra a felis. Integer convallis in orci nec bibendum. Ut consequat posuere metus, www.domain.com.

Curabitur lectus odio, tempus quis velit vitae, cursus sagittis nulla. Maecenas sem nulla, tempor nec risus nec, ultricies ultricies magna. https://127.0.0.1/foo Nulla malesuada lacinia libero non mollis. Curabitur id lacus id dolor vestibulum ornare quis a nisi (http://domain.com/some/path?with=query). Pellentesque ac finibus mauris. Sed eu luctus diam. Quisque porta lectus in turpis imperdiet dapibus.

#blessed #interweave #milesj`}
					matchers={[
						new EmailMatcher('email'),
						new HashtagMatcher('hashtag'),
						new IpMatcher('ip'),
						new UrlMatcher('url'),
					]}
					tagName="div"
				/>,
			);

			expect(implSpy).not.toHaveBeenCalled();
			expect(actual).toBe(
				'<div>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec massa lorem, mollis non commodo quis, ultricies at elit. <a href="mailto:email@domain.com" rel="noopener noreferrer">email@domain.com</a>. Aliquam a arcu porttitor, aliquam eros sed, convallis massa. Nunc vitae vehicula quam, in feugiat ligula. <a href="interweave" rel="noopener noreferrer">#interweave</a> Donec eu sem non nibh condimentum luctus. Vivamus pharetra feugiat blandit. Vestibulum neque velit, semper id vestibulum id, viverra a felis. Integer convallis in orci nec bibendum. Ut consequat posuere metus, <a href="http://www.domain.com" rel="noopener noreferrer">www.domain.com</a>.<br/><br/>Curabitur lectus odio, tempus quis velit vitae, cursus sagittis nulla. Maecenas sem nulla, tempor nec risus nec, ultricies ultricies magna. <a href="https://127.0.0.1/foo" rel="noopener noreferrer">https://127.0.0.1/foo</a> Nulla malesuada lacinia libero non mollis. Curabitur id lacus id dolor vestibulum ornare quis a nisi (<a href="http://domain.com/some/path?with=query" rel="noopener noreferrer">http://domain.com/some/path?with=query</a>). Pellentesque ac finibus mauris. Sed eu luctus diam. Quisque porta lectus in turpis imperdiet dapibus.<br/><br/><a href="blessed" rel="noopener noreferrer">#blessed</a> <a href="interweave" rel="noopener noreferrer">#interweave</a> <a href="milesj" rel="noopener noreferrer">#milesj</a></div>',
			);
		});
	});
});
