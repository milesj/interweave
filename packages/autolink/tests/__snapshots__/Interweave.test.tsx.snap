// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interweave (with autolinking) doesnt render anchor links within anchor links 1`] = `
<Interweave
  content="- https://127.0.0.1/foo
  - <a href="www.domain.com">www.domain.com</a>
  - (http://domain.com/some/path?with=query)
  - <a href="http://domain.com">This text should stay</a>"
  matchers={[ new EmailMatcher(), new HashtagMatcher(), new IpMatcher(), new UrlMatcher() ]}
  tagName="div"
>
  <Markup
    emptyContent={null}
    parsedContent={[
      "- ",
      <Url
      urlParts={{ auth: "", fragment: "", host: "127.0.0.1", path: "/foo", port: "", ... 2 more }}
    />,
      "
    - ",
      <Element attributes={{ href: "www.domain.com" }} tagName="a" />,
      "
    - (",
      ... 3 more,
    ]}
    tagName="div"
  >
    <Element tagName="div">
      <div>
        - 
        <Url
          urlParts={{ auth: "", fragment: "", host: "127.0.0.1", path: "/foo", port: "", ... 2 more }}
        >
          <Link href="https://127.0.0.1/foo">
            <a href="https://127.0.0.1/foo" rel="noopener noreferrer">
              https://127.0.0.1/foo
            </a>
          </Link>
        </Url>
        
- 
        <Element attributes={{ href: "www.domain.com" }} tagName="a">
          <a href="www.domain.com">
            www.domain.com
          </a>
        </Element>
        
- (
        <Url
          urlParts={{
            auth: "",
            fragment: "",
            host: "domain.com",
            path: "/some/path",
            port: "",
            ... 2 more,
          }}
        >
          <Link href="http://domain.com/some/path?with=query">
            <a
              href="http://domain.com/some/path?with=query"
              rel="noopener noreferrer"
            >
              http://domain.com/some/path?with=query
            </a>
          </Link>
        </Url>
        )
- 
        <Element attributes={{ href: "http://domain.com" }} tagName="a">
          <a href="http://domain.com">
            This text should stay
          </a>
        </Element>
      </div>
    </Element>
  </Markup>
</Interweave>
`;

exports[`Interweave (with autolinking) renders HTML text with all matchers 1`] = `
<Interweave
  content="<h1>Lorem ipsum dolor sit amet</h1>
  
  <p><b>Consectetur adipiscing elit.</b> Donec massa lorem, mollis non commodo quis, ultricies at elit. email@domain.com. Aliquam a arcu porttitor, aliquam eros sed, convallis massa. Nunc vitae vehicula quam, in feugiat ligula. #interweave Donec eu sem non nibh condimentum luctus. Vivamus pharetra feugiat blandit. Vestibulum neque velit, semper id vestibulum id, viverra a felis. Integer convallis in orci nec bibendum. Ut consequat posuere metus, <a href="www.domain.com">www.domain.com</a>.</p>
  
  <div>Curabitur lectus odio, <em>tempus quis velit vitae, cursus sagittis nulla</em>. Maecenas sem nulla, tempor nec risus nec, ultricies ultricies magna. https://127.0.0.1/foo Nulla malesuada lacinia libero non mollis. Curabitur id lacus id dolor vestibulum ornare quis a nisi (http://domain.com/some/path?with=query). Pellentesque ac finibus mauris. Sed eu luctus diam. Quisque porta lectus in turpis imperdiet dapibus.</div>
  
  <section>#blessed #interweave #milesj</section>"
  matchers={[ new EmailMatcher(), new HashtagMatcher(), new IpMatcher(), new UrlMatcher() ]}
  tagName="div"
>
  <Markup
    emptyContent={null}
    parsedContent={[
      <Element tagName="h1" />,
      "
    
    ",
      <Element tagName="p" />,
      "
    
    ",
      <Element tagName="div" />,
      ... 2 more,
    ]}
    tagName="div"
  >
    <Element tagName="div">
      <div>
        <Element tagName="h1">
          <h1>Lorem ipsum dolor sit amet</h1>
        </Element>
        


        <Element tagName="p">
          <p>
            <Element tagName="b">
              <b>Consectetur adipiscing elit.</b>
            </Element>
             Donec massa lorem, mollis non commodo quis, ultricies at elit. 
            <Email emailParts={{ host: "domain.com", username: "email" }}>
              <Link href="mailto:email@domain.com">
                <a href="mailto:email@domain.com" rel="noopener noreferrer">
                  email@domain.com
                </a>
              </Link>
            </Email>
            . Aliquam a arcu porttitor, aliquam eros sed, convallis massa. Nunc vitae vehicula quam, in feugiat ligula. 
            <Hashtag hashtagName="interweave">
              <Link hashtagName="interweave" href="interweave">
                <a href="interweave" rel="noopener noreferrer">
                  #interweave
                </a>
              </Link>
            </Hashtag>
             Donec eu sem non nibh condimentum luctus. Vivamus pharetra feugiat blandit. Vestibulum neque velit, semper id vestibulum id, viverra a felis. Integer convallis in orci nec bibendum. Ut consequat posuere metus, 
            <Element attributes={{ href: "www.domain.com" }} tagName="a">
              <a href="www.domain.com">
                www.domain.com
              </a>
            </Element>
            .
          </p>
        </Element>
        


        <Element tagName="div">
          <div>
            Curabitur lectus odio, 
            <Element tagName="em">
              <em>tempus quis velit vitae, cursus sagittis nulla</em>
            </Element>
            . Maecenas sem nulla, tempor nec risus nec, ultricies ultricies magna. 
            <Url
              urlParts={{ auth: "", fragment: "", host: "127.0.0.1", path: "/foo", port: "", ... 2 more }}
            >
              <Link href="https://127.0.0.1/foo">
                <a href="https://127.0.0.1/foo" rel="noopener noreferrer">
                  https://127.0.0.1/foo
                </a>
              </Link>
            </Url>
             Nulla malesuada lacinia libero non mollis. Curabitur id lacus id dolor vestibulum ornare quis a nisi (
            <Url
              urlParts={{
                auth: "",
                fragment: "",
                host: "domain.com",
                path: "/some/path",
                port: "",
                ... 2 more,
              }}
            >
              <Link href="http://domain.com/some/path?with=query">
                <a
                  href="http://domain.com/some/path?with=query"
                  rel="noopener noreferrer"
                >
                  http://domain.com/some/path?with=query
                </a>
              </Link>
            </Url>
            ). Pellentesque ac finibus mauris. Sed eu luctus diam. Quisque porta lectus in turpis imperdiet dapibus.
          </div>
        </Element>
        


        <Element tagName="section">
          <section>
            <Hashtag hashtagName="blessed">
              <Link hashtagName="blessed" href="blessed">
                <a href="blessed" rel="noopener noreferrer">
                  #blessed
                </a>
              </Link>
            </Hashtag>
             
            <Hashtag hashtagName="interweave">
              <Link hashtagName="interweave" href="interweave">
                <a href="interweave" rel="noopener noreferrer">
                  #interweave
                </a>
              </Link>
            </Hashtag>
             
            <Hashtag hashtagName="milesj">
              <Link hashtagName="milesj" href="milesj">
                <a href="milesj" rel="noopener noreferrer">
                  #milesj
                </a>
              </Link>
            </Hashtag>
          </section>
        </Element>
      </div>
    </Element>
  </Markup>
</Interweave>
`;

exports[`Interweave (with autolinking) renders large blocks of text with all matchers 1`] = `
<Interweave
  content="Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec massa lorem, mollis non commodo quis, ultricies at elit. email@domain.com. Aliquam a arcu porttitor, aliquam eros sed, convallis massa. Nunc vitae vehicula quam, in feugiat ligula. #interweave Donec eu sem non nibh condimentum luctus. Vivamus pharetra feugiat blandit. Vestibulum neque velit, semper id vestibulum id, viverra a felis. Integer convallis in orci nec bibendum. Ut consequat posuere metus, www.domain.com.
  
  Curabitur lectus odio, tempus quis velit vitae, cursus sagittis nulla. Maecenas sem nulla, tempor nec risus nec, ultricies ultricies magna. https://127.0.0.1/foo Nulla malesuada lacinia libero non mollis. Curabitur id lacus id dolor vestibulum ornare quis a nisi (http://domain.com/some/path?with=query). Pellentesque ac finibus mauris. Sed eu luctus diam. Quisque porta lectus in turpis imperdiet dapibus.
  
  #blessed #interweave #milesj"
  matchers={[ new EmailMatcher(), new HashtagMatcher(), new IpMatcher(), new UrlMatcher() ]}
  tagName="div"
>
  <Markup
    emptyContent={null}
    parsedContent={[
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec massa lorem, mollis non commodo quis, ultricies at elit. ",
      <Email emailParts={{ host: "domain.com", username: "email" }} />,
      ". Aliquam a arcu porttitor, aliquam eros sed, convallis massa. Nunc vitae vehicula quam, in feugiat ligula. ",
      <Hashtag hashtagName="interweave" />,
      " Donec eu sem non nibh condimentum luctus. Vivamus pharetra feugiat blandit. Vestibulum neque velit, semper id vestibulum id, viverra a felis. Integer convallis in orci nec bibendum. Ut consequat posuere metus, ",
      ... 16 more,
    ]}
    tagName="div"
  >
    <Element tagName="div">
      <div>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec massa lorem, mollis non commodo quis, ultricies at elit. 
        <Email emailParts={{ host: "domain.com", username: "email" }}>
          <Link href="mailto:email@domain.com">
            <a href="mailto:email@domain.com" rel="noopener noreferrer">
              email@domain.com
            </a>
          </Link>
        </Email>
        . Aliquam a arcu porttitor, aliquam eros sed, convallis massa. Nunc vitae vehicula quam, in feugiat ligula. 
        <Hashtag hashtagName="interweave">
          <Link hashtagName="interweave" href="interweave">
            <a href="interweave" rel="noopener noreferrer">
              #interweave
            </a>
          </Link>
        </Hashtag>
         Donec eu sem non nibh condimentum luctus. Vivamus pharetra feugiat blandit. Vestibulum neque velit, semper id vestibulum id, viverra a felis. Integer convallis in orci nec bibendum. Ut consequat posuere metus, 
        <Url
          urlParts={{ auth: "", fragment: "", host: "www.domain.com", path: "", port: "", ... 2 more }}
        >
          <Link href="http://www.domain.com">
            <a href="http://www.domain.com" rel="noopener noreferrer">
              www.domain.com
            </a>
          </Link>
        </Url>
        .
        <Element selfClose tagName="br">
          <br />
        </Element>
        <Element selfClose tagName="br">
          <br />
        </Element>
        Curabitur lectus odio, tempus quis velit vitae, cursus sagittis nulla. Maecenas sem nulla, tempor nec risus nec, ultricies ultricies magna. 
        <Url
          urlParts={{ auth: "", fragment: "", host: "127.0.0.1", path: "/foo", port: "", ... 2 more }}
        >
          <Link href="https://127.0.0.1/foo">
            <a href="https://127.0.0.1/foo" rel="noopener noreferrer">
              https://127.0.0.1/foo
            </a>
          </Link>
        </Url>
         Nulla malesuada lacinia libero non mollis. Curabitur id lacus id dolor vestibulum ornare quis a nisi (
        <Url
          urlParts={{
            auth: "",
            fragment: "",
            host: "domain.com",
            path: "/some/path",
            port: "",
            ... 2 more,
          }}
        >
          <Link href="http://domain.com/some/path?with=query">
            <a
              href="http://domain.com/some/path?with=query"
              rel="noopener noreferrer"
            >
              http://domain.com/some/path?with=query
            </a>
          </Link>
        </Url>
        ). Pellentesque ac finibus mauris. Sed eu luctus diam. Quisque porta lectus in turpis imperdiet dapibus.
        <Element selfClose tagName="br">
          <br />
        </Element>
        <Element selfClose tagName="br">
          <br />
        </Element>
        <Hashtag hashtagName="blessed">
          <Link hashtagName="blessed" href="blessed">
            <a href="blessed" rel="noopener noreferrer">
              #blessed
            </a>
          </Link>
        </Hashtag>
         
        <Hashtag hashtagName="interweave">
          <Link hashtagName="interweave" href="interweave">
            <a href="interweave" rel="noopener noreferrer">
              #interweave
            </a>
          </Link>
        </Hashtag>
         
        <Hashtag hashtagName="milesj">
          <Link hashtagName="milesj" href="milesj">
            <a href="milesj" rel="noopener noreferrer">
              #milesj
            </a>
          </Link>
        </Hashtag>
      </div>
    </Element>
  </Markup>
</Interweave>
`;
